components/de_amp/browser/de_amp_url_loader.cc

```diff
@@ -0,0 +1,184 @@
+
+DeAmpURLLoader::~DeAmpURLLoader() = default;
+
+void DeAmpURLLoader::OnBodyReadable(MojoResult) {
+  if (state_ == State::kSending) {
+    // The pipe becoming readable when kSending means all buffered body has
+    // already been sent.
+    ForwardBodyToClient();
+    return;
+  }
+  if (!BodySnifferURLLoader::CheckBufferedBody(kReadBufferSize)) {
+    return;
+  }
+
+  DeAmp();
+
+  body_consumer_watcher_.ArmOrNotify();
+}
+
+void DeAmpURLLoader::DeAmp() {
+  std::string canonical_link;
+
+  if (de_amp_service->FindCanonicalLinkIfAMP(buffered_body_, &canonical_link)) {
+    const GURL canonical_url(canonical_link);
+    if (!de_amp_service->VerifyCanonicalLink(canonical_url, response_url_)) {
+      VLOG(2) << __func__ << " canonical link check failed " << canonical_url;
+      CompleteLoading(std::move(buffered_body_));
+      return;
+    }
+    VLOG(2) << __func__ << " de-amping and loading " << canonical_url;
+    contents_->GetController().LoadURL(canonical_url, content::Referrer(),
+                                       ui::PAGE_TRANSITION_CLIENT_REDIRECT,
+                                       std::string());
+  } else {
+    // Did not find AMP page and/or canonical link, load original
+    CompleteLoading(std::move(buffered_body_));
+    return;
+  }
+}
+
+void DeAmpURLLoader::OnBodyWritable(MojoResult r) {
+  DCHECK_EQ(State::kSending, state_);
+  if (bytes_remaining_in_buffer_ > 0) {
+    SendReceivedBodyToClient();
+  } else {
+    ForwardBodyToClient();
+  }
+}
+
+void DeAmpURLLoader::ForwardBodyToClient() {
+  DCHECK_EQ(0u, bytes_remaining_in_buffer_);
+  // Send the body from the consumer to the producer.
+  const void* buffer;
+  uint32_t buffer_size = 0;
+  MojoResult result = body_consumer_handle_->BeginReadData(
+      &buffer, &buffer_size, MOJO_BEGIN_READ_DATA_FLAG_NONE);
+  switch (result) {
+    case MOJO_RESULT_OK:
+      break;
+    case MOJO_RESULT_SHOULD_WAIT:
+      body_consumer_watcher_.ArmOrNotify();
+      return;
+    case MOJO_RESULT_FAILED_PRECONDITION:
+      // All data has been sent.
+      CompleteSending();
+      return;
+    default:
+      NOTREACHED();
+      return;
+  }
+
+  result = body_producer_handle_->WriteData(buffer, &buffer_size,
+                                            MOJO_WRITE_DATA_FLAG_NONE);
+  switch (result) {
+    case MOJO_RESULT_OK:
+      break;
+    case MOJO_RESULT_FAILED_PRECONDITION:
+      // The pipe is closed unexpectedly. |this| should be deleted once
+      // URLLoader on the destination is released.
+      BodySnifferURLLoader::Abort();
+      return;
+    case MOJO_RESULT_SHOULD_WAIT:
+      body_consumer_handle_->EndReadData(0);
+      body_producer_watcher_.ArmOrNotify();
+      return;
+    default:
+      NOTREACHED();
+      return;
+  }
+
+  body_consumer_handle_->EndReadData(buffer_size);
+  body_consumer_watcher_.ArmOrNotify();
+}
+
+void DeAmpURLLoader::CompleteSending() {
+  DCHECK_EQ(State::kSending, state_);
+  state_ = State::kCompleted;
```

Comment:
It's best to avoid making superclass properties protected (use methods instead) and also now both subclasses have to know to set it to completed here which is dangerous. The superclass method should call `state_ = State::kCompleted;`